### ISSUE TITLE
[Proposal] Add ReadOnlySpan<char> overloads to Console.Write(Line)

### ISSUE THREAD

**Issue by C-xC-c** [dotnet/runtime#54529]
## Background and Motivation
Currrently, `TextWriter` has overloads for writing `ReadOnlySpan<char>`, but these are not exposed in `Console`. It is possible to write a span to the console via `Console.Out`, but I don't think this is obvious to developers, and the span based overloads should exist on `Console` too.

## Proposed API
```diff
namespace System
{
    public static class Console
    {
+      public static void Write(ReadOnlySpan<char> buffer);
+      public static void WriteLine(ReadOnlySpan<char> buffer);
    }
}
```

## Usage Examples
``` C#
string s = "testing";
Console.Write(s.AsSpan(0,4)); // "test"
```

## Alternative Designs
None.

## Risks
None that I am aware of.

**Comment by GrabYourPitchforks** [comment]
@stephentoub Weren't you thinking of `InterpolatedStringBuilder`-enlightening the `Console` class? Would this naturally precipitate from such work?

**Comment by stephentoub** [comment]
> Weren't you thinking of InterpolatedStringBuilder-enlightening the Console class?

I don't think so? :smile:  Or, if I was, I forgot.

I'm not sure what that would buy in this particular case.  By default with C# 10 and .NET 6, the default support for string interpolation will support `ReadOnlySpan<char>` in holes, so without doing anything special you'll be able to write:
```C#
ReadOnlySpan<char> span = ...;
Console.WriteLine($"{span}");
```
and it'll "just work".  The primary reason we'd go a step further and add a custom interpolated string handler would be if either a) we could do something meaningfully more efficient in the implementation or b) take advantage of conditionality of whether to perform the interpolation or not.  For (a), I'd be surprised if the string allocation were prohibitive, and for (b), I don't think that applies here.  You could imagine adding an implicit conversion from `ROS<char>` to such a handler, in which case you could probably write it as `Console.WriteLine(span)`, but that seems worse than just adding the overload.

But I may not be understanding the question.  Could you elaborate?

As for adding overloads here, personally I'm inclined to say it's fine to either do `Console.Out.WriteLine(span)` or `Console.WriteLine(span.ToString())`, both of which work today.  But I also don't have anything against adding `ReadOnlySpan<char>` overloads if it was really broadly desired.

**Comment by GrabYourPitchforks** [comment]
> But I may not be understanding the question. Could you elaborate?

I think your response pretty much nailed what I was trying to ask. :)

From looking at the `Console.WriteLine` implementation, it looks like a wrapper around a non-subclassed `StreamWriter` with some additional ceremony, and `StreamWriter` appears to have an optimized `WriteLine(ROS<char>)` implementation. So at first glance this could be plumbed through. But this is all based on a very brief glance.

**Comment by stephentoub** [comment]
> So at first glance this could be plumbed through.

Yes, _if_ we decided to add the API, the implementation would just be:
```C#
public static void WriteLine(ReadOnlySpan<char> value) => Out.WriteLine(value);
```
:smile:

**Comment by GSPP** [comment]
I wonder if saving this string allocation would be meaningful given how expensive it is to write to the console in the first place. I'd suggest benchmarking this first before adding an API.

**Comment by stephentoub** [comment]
Given https://github.com/dotnet/runtime/issues/54529#issuecomment-865434595, I'm going to close the issue.  Thank you for the suggestion.
